<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [2]:
# Import pandas and create_engine from `sqlalchemy`

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data
engine = \
create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT( DISTINCT store)
FROM stores

"""

store = pd.read_sql_query(sql, engine)

In [7]:
store 

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [10]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT name
FROM stores
ORDER BY name
LIMIT 5
"""

pd.read_sql_query(sql, engine)

name
0  218 Fuel Express & Chubby's Liquor
1          3047 Cub Foods / Iowa City
2               3060 Cub Foods / Ames
3         3061 Cub Foods / Sioux City
4                  3rd St Convenience

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [22]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, btl_price, bottle_qty
FROM sales
WHERE category_name = 'IMPORTED VODKA - MISC'
ORDER BY vendor
LIMIT 3;
"""

pd.read_sql_query(sql, engine)

vendor btl_price  bottle_qty
0  Bacardi U.S.A. Inc.    $27.74           1
1  Bacardi U.S.A. Inc.    $27.74           1
2  Bacardi U.S.A. Inc.    $27.74           2

**Prompt 3 Answer**: 
	vendor	btl_price	bottle_qty
0	Bacardi U.S.A. Inc.	$27.74	1
1	Bacardi U.S.A. Inc.	$27.74	1
2	Bacardi U.S.A. Inc.	$27.74	2


### Prompt 4

What is the total amount of liquor sold in February 2015?

In [24]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT DATE_TRUNC('month', date) as sales_month,
SUM(total) AS monthly_sales
FROM sales
GROUP BY 1
"""
pd.read_sql_query(sql, engine)

sales_month  monthly_sales
0   2014-01-01    40331023.96
1   2014-02-01    38543588.42
2   2014-03-01    38603543.82
3   2014-04-01    45388494.64
4   2014-05-01    42110068.64
5   2014-06-01    24414597.17
6   2014-07-01    22096736.93
7   2014-08-01    21207347.84
8   2014-09-01    23440959.93
9   2014-10-01    26752602.20
10  2014-11-01    20244210.54
11  2014-12-01     9306222.86
12  2015-01-01    18558276.11
13  2015-02-01    21295350.55

**Prompt 4 Answer**: 21295350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [52]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, COUNT(*) as total_transaction, SUM(bottle_qty) as total_quantity, SUM(total) as total_revenue
FROM sales
WHERE category_name LIKE '%%IMPORTED VODKA%%'
GROUP BY vendor
ORDER BY total_revenue DESC
LIMIT 3;
"""

pd.read_sql_query(sql, engine)

vendor  total_transaction  total_quantity  \
0  Pernod Ricard USA/Austin Nichols              67190          570679   
1               Bacardi U.S.A. Inc.              29054          244756   
2                   Diageo Americas              31672          281482   

   total_revenue  
0    10902960.40  
1     6006840.37  
2     5699011.89

**Prompt 5 Answer**: 
          vendor	                            total_transaction	total_quantity	total_revenue
0	Pernod Ricard USA/Austin Nichols	67190	              570679	    10902960.40

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [27]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT btl_price - state_btl_cost as Markup
FROM sales
"""

pd.read_sql_query(sql, engine)

markup
0         $5.75
1         $5.75
2         $5.75
3         $8.58
4         $2.50
...         ...
3049908   $5.75
3049909  $10.00
3049910   $5.75
3049911   $5.75
3049912  $10.00

[3049913 rows x 1 columns]

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [24]:
sql = """
SELECT store,name, ROUND(AVG(Percentage_Markup), 2) as avg_Percentage_Markup
FROM(SELECT s.store, name, 100.0*(btl_price - state_btl_cost)::NUMERIC/state_btl_cost::NUMERIC as Percentage_Markup
FROM sales as s
JOIN stores as st
ON s.store =st.store) as temp
GROUP BY store, name
ORDER BY avg_Percentage_Markup DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

store                        name  avg_percentage_markup
0   4013      Wal-Mart 0841 / Tipton                  52.55
1   4024  Wal-Mart 1546 / Iowa Falls                  52.15
2   4266  Wal-Mart 1683 / Shenandoah                  51.89
3   3833    Wal-Mart 3394 / Atlantic                  51.51
4   3660   Wal-Mart 2935 / Knoxville                  51.46

**Prompt 7 Answer**: 
Wal-Mart 0841 / Tipton	with 52.547124% mark up

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [25]:
sql = """
SELECT *
FROM sales

LIMIT 5;
"""

pd.read_sql_query(sql, engine)

date convenience_store  store county_number county category  \
0 2014-11-13              None   2635            82  Scott  1032200   
1 2014-11-13              None   2635            82  Scott  1032200   
2 2014-11-13              None   2635            82  Scott  1032200   
3 2014-11-13              None   2635            82  Scott  1032080   
4 2014-11-13              None   2635            82  Scott  1032080   

           category_name vendor_no                            vendor   item  \
0  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34036   
1  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34076   
2  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34116   
3         IMPORTED VODKA       420            Moet Hennessy USA Inc.  34155   
4         IMPORTED VODKA       260                   Diageo Americas  34164   

          description  pack  liter_size state_btl_cost btl_price  bottle_qty  \
0       Absolut Pears    12         750         $11.49    $17.24           4   
1     Absolut Vanilia    12         750         $11.49    $17.24          12   
2     Absolut Mandrin    12         750         $11.49    $17.24          12   
3     Belvedere Vodka     6         750         $17.15    $25.73           6   
4  Ciroc Luxury Vodka    24         200          $5.00     $7.50           6   

    total  
0   68.96  
1  206.88  
2  206.88  
3  154.38  
4   45.00

In [43]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store, name, store_address, 
ROUND(AVG(Percentage_Markup), 2) as avg_Percentage_Markup, 
ROUND(MIN(Percentage_Markup), 2) as min_mark_up,
ROUND(MAX(Percentage_Markup), 2) as max_mark_up
FROM(SELECT s.store, name, store_address,
100.0*(btl_price - state_btl_cost)::NUMERIC/state_btl_cost::NUMERIC as Percentage_Markup
FROM sales as s
JOIN stores as st
ON s.store =st.store) as temp
GROUP BY store, name, store_address
ORDER BY avg_Percentage_Markup DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

store                        name  \
0   4013      Wal-Mart 0841 / Tipton   
1   4024  Wal-Mart 1546 / Iowa Falls   
2   4266  Wal-Mart 1683 / Shenandoah   
3   3833    Wal-Mart 3394 / Atlantic   
4   3660   Wal-Mart 2935 / Knoxville   

                                       store_address  avg_percentage_markup  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...                  52.55   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...                  52.15   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...                  51.89   
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....                  51.51   
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...                  51.46   

   min_mark_up  max_mark_up  
0        49.86       670.88  
1        49.85       670.88  
2        49.85       670.88  
3        49.85       670.88  
4        49.85       670.88

**Prompt 8 Answer**:
      store  	name	store_address	                                                                        avg_percentage_markup	min_mark_up	    max_mark_up
1	4013	Wal-Mart 0841 / Tipton	1126 Highway 38 North\nTipton, IA 527720000\n(...	        52.55	                        49.86	               670.88
2	4024	Wal-Mart 1546 / Iowa Falls	840 S Oak\nIowa Falls, IA 501260000\n(42.50295...	   52.15	                        49.85	               670.88
3	4266	Wal-Mart 1683 / Shenandoah	705 S Fremont\nShenandoah, IA 516010000\n(40.7...51.89	                          49.85	                 670.88
4	3833	Wal-Mart 3394 / Atlantic	1905 East 7th St\nAtlantic, IA 500220000\n(41....	        51.51	                         49.85	                670.88
5	3660	Wal-Mart 2935 / Knoxville	814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...	     51.46	                          49.85	                  670.88

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery).

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT county, ROUND(total_county_bottle/population, 2) as total_bottle_percapita
FROM( SELECT s.county, population, SUM(bottle_qty) as total_county_bottle
FROM sales as s
JOIN counties as c
ON s.county = c.county
GROUP BY s.county, population) as temp
ORDER BY total_bottle_percapita DESC
LIMIT 9
"""

pd.read_sql_query(sql, engine)

county  total_bottle_percapita
0      Dickinson                    22.0
1           Polk                    15.0
2     Black Hawk                    15.0
3    Cerro Gordo                    14.0
4          Scott                    13.0
5        Johnson                    13.0
6           Linn                    12.0
7  Pottawattamie                    11.0
8        Kossuth                    11.0

**Prompt 9 Answer**:
county	total_bottle_percapita
0	Dickinson	22.0
1	Polk	15.0
2	Black Hawk	15.0
3	Cerro Gordo	14.0
4	Scott	13.0
5	Johnson	13.0
6	Linn	12.0
7	Pottawattamie	11.0
8	Kossuth	11.0

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [17]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*)
FROM( SELECT total::NUMERIC - (btl_price*bottle_qty)::NUMERIC as diff_total
FROM sales) AS temp
WHERE diff_total !=0
"""

pd.read_sql_query(sql, engine)

count
0      0

In [23]:
sql = """
SELECT total::NUMERIC - (btl_price*bottle_qty)::NUMERIC as diff_total
FROM sales
ORDER BY diff_total DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

diff_total
0         0.0
1         0.0
2         0.0
3         0.0
4         0.0

In [26]:
sql = """
SELECT total, (btl_price*bottle_qty)::NUMERIC as cal_total
FROM sales
WHERE total != (btl_price*bottle_qty)::NUMERIC
"""

pd.read_sql_query(sql, engine)

Empty DataFrame
Columns: [total, cal_total]
Index: []

**Prompt 10 Answer**:
0

Prompt 11 : How many total products are in the Products table?

In [28]:
sql = """
SELECT COUNT(DISTINCT item_no)
FROM products

"""

pd.read_sql_query(sql, engine)

count
0   9977

**Prompt 11 Answer**: 
    9977

Prompt 12 : Who are the top most diverse vendors (i.e. they have the highest number of distinct products)? How many different products do they have?

In [32]:
sql = """
SELECT vendor, vendor_name, COUNT(DISTINCT item_no) as product_no
FROM products
GROUP BY vendor, vendor_name
ORDER BY product_no DESC
LIMIT 3
"""

pd.read_sql_query(sql, engine)

vendor                       vendor_name  product_no
0      65                   Jim Beam Brands         925
1     260                   Diageo Americas         907
2     370  Pernod Ricard Usa/austin Nichols         599

**Prompt 12 Answer**:  
Jim Beam Brands with 925 products

Prompt 13: Which products sell the best by total number of unit sales? (Hint: you will need to find a way of doing a cross-join between the sales and product tables.)

In [36]:
sql = """
SELECT s.item, item_description, SUM(bottle_qty) as total_bottle
FROM sales as s
JOIN products as p
ON s.item = p.item_no
GROUP BY s.item, item_description
ORDER BY SUM(bottle_qty) DESC
LIMIT 3
"""

pd.read_sql_query(sql, engine)

item           item_description  total_bottle
0  11788               Black Velvet        828393
1  36308              Hawkeye Vodka        500070
2  43337  Captain Morgan Spiced Rum        483621

**Prompt 13 Answer**: Black Velvet was sold by 828393 bottles

**Prompt 14 : Which products sell the best by total dollar value of sales?**

In [37]:
sql = """
SELECT s.item, item_description, SUM(total) as total_sales
FROM sales as s
JOIN products as p
ON s.item = p.item_no
GROUP BY s.item, item_description
ORDER BY total_sales DESC
LIMIT 3
"""

pd.read_sql_query(sql, engine)

item               item_description  total_sales
0  11788                   Black Velvet  12863376.81
1  43337      Captain Morgan Spiced Rum   8523370.55
2  26827  Jack Daniels Old #7 Black Lbl   6668863.94

**Prompt 14 Answer**: Black Velvet was sold for $12863376.81

**Prompt 15 : What are the top 10 categories of liquor sold based on the total amount of sales revenue?**

In [42]:
sql = """
SELECT s.category_name, SUM(total) as total_sales
FROM sales as s
JOIN products as p
ON s.item = p.item_no
GROUP BY s.category_name
ORDER BY total_sales DESC
LIMIT 10
"""

pd.read_sql_query(sql, engine)

category_name  total_sales
0                 CANADIAN WHISKIES  48053061.91
1                    80 PROOF VODKA  48045034.70
2                        SPICED RUM  31600618.50
3                    IMPORTED VODKA  23879524.63
4                           TEQUILA  21409700.64
5         STRAIGHT BOURBON WHISKIES  20904099.16
6                   WHISKEY LIQUEUR  19338225.94
7                TENNESSEE WHISKIES  17602242.30
8  PUERTO RICO & VIRGIN ISLANDS RUM  12729072.76
9                  BLENDED WHISKIES  12018198.55

**Prompt 16 : Which rum products have sales greater than $10,000? How about whiskey or vodka products??**

In [49]:
sql = """
SELECT s.category_name, SUM(total) as total_sales
FROM sales as s
JOIN products as p
ON s.item = p.item_no
WHERE s.category_name LIKE '%%RUM%%'
GROUP BY s.category_name
HAVING SUM(total) > 10000
ORDER BY total_sales DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

category_name  total_sales
0                        SPICED RUM  31600618.50
1  PUERTO RICO & VIRGIN ISLANDS RUM  12729072.76
2                      FLAVORED RUM   8011699.50
3                      BARBADOS RUM    396024.49
4                       JAMAICA RUM    371404.44

In [50]:
sql = """
SELECT s.category_name, SUM(total) as total_sales
FROM sales as s
JOIN products as p
ON s.item = p.item_no
WHERE s.category_name LIKE '%%WHISKEY%%'
GROUP BY s.category_name
HAVING SUM(total) > 10000
ORDER BY total_sales DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

category_name  total_sales
0  WHISKEY LIQUEUR  19338225.94

In [51]:
sql = """
SELECT s.category_name, SUM(total) as total_sales
FROM sales as s
JOIN products as p
ON s.item = p.item_no
WHERE s.category_name LIKE '%%VODKA%%'
GROUP BY s.category_name
HAVING SUM(total) > 10000
ORDER BY total_sales DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

category_name  total_sales
0         80 PROOF VODKA  48045034.70
1         IMPORTED VODKA  23879524.63
2         FLAVORED VODKA  11537418.94
3  IMPORTED VODKA - MISC   9072915.58
4        100 PROOF VODKA    319636.81

**Prompt 16 Answer**: 
    category_name	total_sales
SPICED RUM	31600618.50
WHISKEY LIQUEUR	19338225.94
80 PROOF VODKA	48045034.70

**Prompt 17 : Which county sold the most amount of vodka during February 2014?**

In [70]:
sql = """

SELECT s.county, DATE_TRUNC('month', date) AS sales_month,
SUM(total) AS monthly_sales
FROM sales AS s
JOIN counties AS c
ON s.county = c.county
WHERE DATE_TRUNC('month', date) = '2014-02-01' AND s.category_name LIKE '%%VODKA%%'
GROUP BY s.county, sales_month
ORDER BY monthly_sales DESC
LIMIT 5

"""

pd.read_sql_query(sql, engine)

county sales_month  monthly_sales
0        Polk  2014-02-01     2259389.14
1        Linn  2014-02-01      737865.34
2       Scott  2014-02-01      688957.32
3     Johnson  2014-02-01      680700.04
4  Black Hawk  2014-02-01      561431.16

**Prompt 17 Answer**: Polk county with total vodka sales as $2259389.14

**Prompt 18 : Which counties were in the top 10 counties for vodka sales in any month in 2014??**

In [91]:
sql = """
SELECT*
FROM(SELECT *,
RANK() OVER (PARTITION BY sales_month
ORDER BY monthly_sales DESC)
FROM (SELECT DATE_TRUNC('month', date) AS sales_month, s.county,
SUM(total) AS monthly_sales
FROM sales AS s
JOIN counties AS c
ON s.county = c.county
WHERE s.category_name LIKE '%%VODKA%%' AND DATE_PART('year', date) = 2014
GROUP BY s.county, sales_month) as temp) as temp2
WHERE rank <11


"""

pd.read_sql_query(sql, engine)

sales_month         county  monthly_sales  rank
0    2014-01-01           Polk     2188692.56     1
1    2014-01-01           Linn      975547.22     2
2    2014-01-01          Scott      764583.12     3
3    2014-01-01        Johnson      687857.74     4
4    2014-01-01     Black Hawk      611497.16     5
..          ...            ...            ...   ...
115  2014-12-01  Pottawattamie       55802.73     6
116  2014-12-01     Des Moines       46914.00     7
117  2014-12-01        Dubuque       44585.21     8
118  2014-12-01    Cerro Gordo       38714.99     9
119  2014-12-01          Story       38213.19    10

[120 rows x 4 columns]

**Prompt 19**:  Create a report that shows how many times a county appeared in the “top 10 counties for vodka sales in a month” list over the course of 2014?

In [94]:
sql = """
SELECT county, COUNT(*) AS times_in_top_ten
FROM(SELECT *,
RANK() OVER (PARTITION BY sales_month
ORDER BY monthly_sales DESC)
FROM (SELECT DATE_TRUNC('month', date) AS sales_month, s.county,
SUM(total) AS monthly_sales
FROM sales AS s
JOIN counties AS c
ON s.county = c.county
WHERE s.category_name LIKE '%%VODKA%%' AND DATE_PART('year', date) = 2014
GROUP BY s.county, sales_month) as temp) as temp2
WHERE rank <11
GROUP BY county
ORDER BY times_in_top_ten DESC
"""

pd.read_sql_query(sql, engine)

county  times_in_top_ten
0      Black Hawk                12
1         Dubuque                12
2         Johnson                12
3            Linn                12
4            Polk                12
5   Pottawattamie                12
6           Scott                12
7        Woodbury                11
8           Story                10
9     Cerro Gordo                 7
10         Dallas                 6
11     Des Moines                 1
12      Dickinson                 1

**Prompt 20**: What is the trend of sales by month? Break up variables such as bottle_price or liter_size into categories (for example: cheap, medium, or expensive).

**Prompt 21**:Which stores sell one of the top five most expensive bottles of alcohol?

In [5]:
sql = """
SELECT DISTINCT name as store_sell_top_five_expensive
FROM(
SELECT name, btl_price, 
DENSE_RANK() OVER (ORDER BY btl_price DESC)
FROM sales as s
JOIN stores as st
ON s.store =st.store
ORDER BY btl_price DESC) AS temp
WHERE dense_rank <6
"""

pd.read_sql_query(sql, engine)

store_sell_top_five_expensive
0                      Benz Distributing
1                         Central City 2
2              Central City Liquor, Inc.
3            Charlie's Wine and Spirits,
4              Grand Falls Casino Resort
5           Hy-vee #3 / Bdi / Des Moines
6   Hy-Vee Food and Drug #6 / Cedar Rapi
7    Hy-vee Food Store #5 / Cedar Rapids
8                     Hy-Vee / Urbandale
9     Hy-Vee Wine and Spirits / Waterloo
10                        John's Grocery
11           Oelwein Bottle and Can Inc.
12                  Quicker Liquor Store
13             Sam's Club 4973 / Dubuque
14     Sam's Club 6344 / Windsor Heights
15          Sam's Club 6432 / Sioux City
16      Sam's Club 6472 / Council Bluffs
17            Sam's Club 6514 / Waterloo
18                Sam's Club 6568 / Ames
19        Sam's Club 8162 / Cedar Rapids
20           Sam's Club 8238 / Davenport

**Prompt 22**:How many stores have more than $2,000,000 in total sales?

In [12]:
sql = """
SELECT COUNT(*)
FROM(
SELECT s.store, SUM(total) as total_sale
FROM sales as s
JOIN stores as st
ON s.store =st.store
GROUP BY s.store
ORDER BY total_sale DESC) AS temp
WHERE total_sale>2000000
"""
pd.read_sql_query(sql, engine)

count
0     24

**Prompt 22 Answer** 24 stores

**Prompt 23**:How many stores have an average bottle price greater than $20?

In [16]:
sql = """
SELECT COUNT(*)
FROM(
SELECT s.store, AVG(btl_price::NUMERIC) as avg_btl_price
FROM sales as s
JOIN stores as st
ON s.store =st.store
GROUP BY s.store
ORDER BY avg_btl_price DESC) AS temp
WHERE avg_btl_price >20
"""
pd.read_sql_query(sql, engine)

count
0     21

**Prompt 23 Answer** 21 stores

**Prompt 24** :Which stores have the highest sales of items over 90 proof? 

In [28]:
sql = """
SELECT s.store,name, SUM(total) as total_sale
FROM sales as s
JOIN products as p
ON s.item = p.item_no
JOIN stores as st
ON s.store =st.store
WHERE proof::NUMERIC > 90
GROUP BY s.store, name
ORDER BY total_sale DESC
LIMIT 5
"""
pd.read_sql_query(sql, engine)

store                                 name  total_sale
0   2633         Hy-vee #3 / Bdi / Des Moines   689442.44
1   4829                       Central City 2   564993.84
2   2512  Hy-vee Wine and Spirits / Iowa City   361513.53
3   3814                Costco Wholesale #788   334316.94
4   3385       Sam's Club 8162 / Cedar Rapids   285790.98

In [25]:
sql = """
SELECT *
FROM products
LIMIT 5
"""
pd.read_sql_query(sql, engine)

item_no                       category_name           item_description  \
0   904616  MISC. AMERICAN CORDIALS & LIQUEURS     Travis Hasse Apple Pie   
1   904617  MISC. IMPORTED CORDIALS & LIQUEURS         D'aristi Xtabentun   
2   904618                      PEACH BRANDIES  Hiram Walker Peach Brandy   
3   904619                     SCOTCH WHISKIES           Oak Cross Whisky   
4   904620                      FLAVORED VODKA       Uv Red(cherry) Vodka   

   vendor                         vendor_name  bottle_size  pack  inner_pack  \
0     305                             Mhw Ltd          750    12           1   
1     391  Anchor Distilling (preiss Imports)          750    12           1   
2     370    Pernod Ricard Usa/austin Nichols         1000    12           1   
3     305                             Mhw Ltd          750     6           1   
4     380           Phillips Beverage Company          200    24           1   

    age proof  list_date   upc             scc bottle_price  shelf_price  \
0  None    40 2009-02-11  None            None        $9.77        14.66   
1  None    60 2009-04-24  None            None       $14.12        21.18   
2  None    60 2009-12-16  None     89540500867        $6.50         9.74   
3  None    86 2009-12-24  None            None       $25.33        38.00   
4  None    60 2009-11-05  None  10087100000000        $1.97         2.96   

   case_cost  
0     117.22  
1     169.40  
2      77.94  
3     152.00  
4      47.30

**Prompt 24 Answer** Hy-vee #3 / Bdi / Des Moines with total sale as	689442.44

**Prompt 25:** Store 2238 (Adventureland Inn at 3200 Adventureland Dr) sold $883.24 in April and $27,526.38 in May, for a 3017% growth rate. That was the highest percentage month-on-month growth rate. Create a query that shows this and the next 9 highest after that.

In [49]:
sql = """
SELECT *, ROUND(100*(monthly_sales-monthly_sales_lag1)/monthly_sales_lag1, 0) as monthly_growth 
FROM(
SELECT *,
LAG(monthly_sales, 1) OVER (PARTITION BY store) AS monthly_sales_lag1
FROM(
SELECT *,
ROW_NUMBER() OVER 
(PARTITION BY store ORDER BY sales_month)
FROM(
SELECT DATE_TRUNC('month', date) as sales_month, s.store, name, store_address,
SUM(total) AS monthly_sales
FROM sales as s
JOIN stores as st
ON s.store =st.store
GROUP BY sales_month, s.store, name, store_address) as temp) as temp2) as temp3
WHERE monthly_sales_lag1 IS NOT NULL
ORDER BY monthly_growth DESC
LIMIT 10
"""
pd.read_sql_query(sql, engine)

sales_month  store                               name  \
0  2014-05-01   2238                  Adventureland Inn   
1  2015-01-01   5050     Fareway Stores #147 / Carlisle   
2  2014-09-01   4275    Fareway Stores #980 / Knoxville   
3  2014-11-01   4169            Super Quick 2 / Hubbell   
4  2015-01-01   4255  FAREWAY STORES #058 / ORANGE CITY   
5  2014-10-01   4669              Vom Fass / Des Moines   
6  2014-03-01   3866         Target Store T-1170 / Ames   
7  2014-08-01   4717      Walgreens #05512 / Bettendorf   
8  2015-01-01   4104   Fareway Stores #491 / Mason City   
9  2015-01-01   4248      Fareway Stores #021 / Sheldon   

                                       store_address  monthly_sales  \
0  3200 Adventureland Dr\nAltoona, IA 500090000\n...       27526.38   
1  1115, Bluestem Dr\nCarlisle, IA 50047\n(41.497...       26055.32   
2  1308 S Lincoln St\nKnoxville, IA 501380000\n(4...       41674.59   
3  1824 Hubbell Ave\nDes Moines, IA 503170000\n(4...       12794.70   
4  512 8th Se\nOrange City, IA 510410000\n(42.997...       20663.34   
5  833, 42nd St\nDes Moines, IA 50312\n(41.593280...        7231.01   
6  320 S Duff\nAmes, IA 500100000\n(42.0193409080...       35800.68   
7  3425 Middle Rd\nBettendorf, IA 52722\n(41.5508...       14064.95   
8  400 North Delaware Ave\nMason City, IA 5040100...       20584.26   
9  2603 Park St\nSheldon, IA 512010000\n(43.18608...        7799.22   

   row_number  monthly_sales_lag1  monthly_growth  
0           5              883.24          3017.0  
1           3             1696.54          1436.0  
2           9             2925.30          1325.0  
3           9              985.92          1198.0  
4          13             1598.77          1192.0  
5           8              649.24          1014.0  
6           3             3301.92           984.0  
7           7             1412.70           896.0  
8          13             2150.64           857.0  
9          13              837.18           832.0